src/psbt.c

```diff
@@ -0,0 +1,2218 @@
+
+    if (!(result_partial_sigs = clone_partial_sigs_map(partial_sigs))) {
+        return WALLY_ENOMEM;
+    }
+    if (input->partial_sigs) {
+        wally_partial_sigs_map_free(input->partial_sigs);
+    }
+    input->partial_sigs = result_partial_sigs;
+    return WALLY_OK;
+}
+
+int wally_psbt_input_set_unknowns(
+    struct wally_psbt_input *input,
+    struct wally_unknowns_map *unknowns)
+{
+    struct wally_unknowns_map *result_unknowns;
+
+    if (!(result_unknowns = clone_unknowns_map(unknowns))) {
+        return WALLY_ENOMEM;
+    }
+    if (input->unknowns) {
+        wally_unknowns_map_free(input->unknowns);
+    }
+    input->unknowns = result_unknowns;
+    return WALLY_OK;
+}
+
+int wally_psbt_input_set_sighash_type(
+    struct wally_psbt_input *input,
+    uint32_t sighash_type)
+{
+    input->sighash_type = sighash_type;
+    return WALLY_OK;
+}
+
+int wally_psbt_input_free(struct wally_psbt_input *input)
+{
+    if (input) {
+        wally_tx_free(input->non_witness_utxo);
+        wally_tx_output_free(input->witness_utxo);
+        clear_and_free(input->redeem_script, input->redeem_script_len);
+        clear_and_free(input->witness_script, input->witness_script_len);
+        clear_and_free(input->final_script_sig, input->final_script_sig_len);
+        wally_tx_witness_stack_free(input->final_witness);
+        wally_keypath_map_free(input->keypaths);
+        wally_partial_sigs_map_free(input->partial_sigs);
+        wally_unknowns_map_free(input->unknowns);
+    }
+    return WALLY_OK;
+}
+
+int wally_psbt_output_init_alloc(
+    unsigned char *redeem_script,
+    size_t redeem_script_len,
+    unsigned char *witness_script,
+    size_t witness_script_len,
+    struct wally_keypath_map *keypaths,
+    struct wally_unknowns_map *unknowns,
+    struct wally_psbt_output **output)
+{
+    struct wally_psbt_output *result;
+    int ret = WALLY_OK;
+
+    TX_CHECK_OUTPUT;
+    TX_OUTPUT_ALLOC(struct wally_psbt_output);
+
+    if (redeem_script && (ret = wally_psbt_output_set_redeem_script(result, redeem_script, redeem_script_len)) != WALLY_OK) {
+        goto fail;
+    }
+    if (witness_script && (ret = wally_psbt_output_set_witness_script(result, witness_script, witness_script_len)) != WALLY_OK) {
+        goto fail;
+    }
+    if (keypaths && (ret = wally_psbt_output_set_keypaths(result, keypaths)) != WALLY_OK) {
+        goto fail;
+    }
+    if (unknowns && (ret = wally_psbt_output_set_unknowns(result, unknowns)) != WALLY_OK) {
+        goto fail;
+    }
+
+    return ret;
+
+fail:
+    wally_psbt_output_free(result);
+    *output = NULL;
+    return ret;
+}
+
+int wally_psbt_output_set_redeem_script(
+    struct wally_psbt_output *output,
+    unsigned char *redeem_script,
+    size_t redeem_script_len)
+{
+    unsigned char *result_redeem_script;
+
+    if (!clone_bytes(&result_redeem_script, redeem_script, redeem_script_len)) {
+        return WALLY_ENOMEM;
+    }
+    if (output->redeem_script) {
```

Comment:
this if shouldn't be needed if the initialisation is OK. same for the others. nit: could possibly use the same setup as the TX_SET_B in transaction.c

Reply:
How so? `output->redeem_script` could always be empty, and AFAICT `wally_free` doesn't check for NULL before freeing.

Reply:
wally_free should forward to free which is a no op for null ptr

Reply:
Done